File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Deque;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;


public class Main {
    public static void main(String[] args) throws IOException {
        //초기화
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int max = Integer.MIN_VALUE;
        int[][] arr = new int[n + 3][m + 3];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            for (int j = 0; j < m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
            for (int j = m; j < m + 3; j++) {
                arr[i][j] = -9_999_999;
            }
        }
        for (int i = n; i < n + 3; i++) {
            for (int j = 0; j < m + 3; j++) {
                arr[i][j] = -9_999_999;
            }
        }
        int[][][] tetriminos = {
                {{1, 0, 0, 0}, {1, 1, 0, 0}, {0, 1, 0, 0}, {0, 0, 0, 0}},
                {{0, 1, 1, 0}, {1, 1, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{0, 1, 0, 0}, {1, 1, 0, 0}, {1, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 0, 0}, {0, 1, 1, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 1, 0}, {0, 1, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{0, 1, 0, 0}, {1, 1, 0, 0}, {0, 1, 0, 0}, {0, 0, 0, 0}},
                {{0, 1, 0, 0}, {1, 1, 1, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 0, 0, 0}, {1, 1, 0, 0}, {1, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 1, 1}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 0, 0, 0}, {1, 0, 0, 0}, {1, 0, 0, 0}, {1, 0, 0, 0}},
                {{1, 1, 0, 0}, {1, 1, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 0, 0, 0}, {1, 0, 0, 0}, {1, 1, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 1, 0}, {1, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 0, 0}, {0, 1, 0, 0}, {0, 1, 0, 0}, {0, 0, 0, 0}},
                {{0, 0, 1, 0}, {1, 1, 1, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{0, 1, 0, 0}, {0, 1, 0, 0}, {1, 1, 0, 0}, {0, 0, 0, 0}},
                {{1, 0, 0, 0}, {1, 1, 1, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 0, 0}, {1, 0, 0, 0}, {1, 0, 0, 0}, {0, 0, 0, 0}},
                {{1, 1, 1, 0}, {0, 0, 1, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}}
        };

        for (int i = 0; i < 19; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < m; k++) {
                    int tmp = 0;
                    for (int l = 0; l < 4; l++) {
                        for (int o = 0; o < 4; o++) {
                            tmp += arr[j + l][k + o] * tetriminos[i][l][o];
                        }
                    }
                    if (max < tmp) {
                        max = tmp;
                    }
                }

            }
        }
        System.out.println(max);
    }
}

File: 백준/Gold/14500. 테트로미노/테트로미노.kt
import kotlin.Throws
import java.io.IOException
import kotlin.jvm.JvmStatic
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

object Main {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        //초기화
        val br = BufferedReader(InputStreamReader(System.`in`))
        var st = StringTokenizer(br.readLine(), " ")
        val n = st.nextToken().toInt()
        val m = st.nextToken().toInt()
        var max = Int.MIN_VALUE
        val arr = Array(n + 3) { IntArray(m + 3) }
        for (i in 0 until n) {
            st = StringTokenizer(br.readLine(), " ")
            for (j in 0 until m) {
                arr[i][j] = st.nextToken().toInt()
            }
            for (j in m until m + 3) {
                arr[i][j] = -9999999
            }
        }
        for (i in n until n + 3) {
            for (j in 0 until m + 3) {
                arr[i][j] = -9999999
            }
        }
        val tetriminos = arrayOf(
            arrayOf(intArrayOf(1, 0, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(0, 1, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(0, 1, 1, 0), intArrayOf(1, 1, 0, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(0, 1, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 0, 0), intArrayOf(0, 1, 1, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 1, 0), intArrayOf(0, 1, 0, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(0, 1, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(0, 1, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(0, 1, 0, 0), intArrayOf(1, 1, 1, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 0, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 1, 1), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 0, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(1, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 0, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 1, 0), intArrayOf(1, 0, 0, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 0, 0), intArrayOf(0, 1, 0, 0), intArrayOf(0, 1, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(0, 0, 1, 0), intArrayOf(1, 1, 1, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(0, 1, 0, 0), intArrayOf(0, 1, 0, 0), intArrayOf(1, 1, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 0, 0, 0), intArrayOf(1, 1, 1, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(0, 0, 0, 0)),
            arrayOf(intArrayOf(1, 1, 1, 0), intArrayOf(0, 0, 1, 0), intArrayOf(0, 0, 0, 0), intArrayOf(0, 0, 0, 0))
        )
        for (i in 0..18) {
            for (j in 0 until n) {
                for (k in 0 until m) {
                    var tmp = 0
                    for (l in 0..3) {
                        for (o in 0..3) {
                            tmp += arr[j + l][k + o] * tetriminos[i][l][o]
                        }
                    }
                    if (max < tmp) {
                        max = tmp
                    }
                }
            }
        }
        println(max)
    }
}